src/FWlib/peripheral_keyscan.c

```diff
@@ -0,0 +1,284 @@
+    uint8_t offset = 29;
+    uint8_t bits_width = 1;
+    uint32_t data = (enable == 0) ? 0: 1;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_scanner_ctrl0, offset, bits_width, data);
+    return;
+}
+
+void KEYSCAN_DbClkSel(uint8_t sel)
+{
+    uint8_t offset = 30;
+    uint8_t bits_width = 1;
+    uint32_t data = (sel == 0) ? 0: 1;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_scanner_ctrl0, offset, bits_width, data);
+    return;
+}
+
+void KEYSCAN_SetReleaseTime(uint32_t time)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 11;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_scanner_ctrl1, offset, bits_width, time);
+    return;
+}
+
+void KEYSCAN_SetScanInterval(uint32_t scan_itv)
+{
+    uint8_t offset = 11;
+    uint8_t bits_width = 16;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_scanner_ctrl1, offset, bits_width, scan_itv);
+    return;
+}
+
+void KEYSCAN_SetOutRowMask(uint32_t row_mask)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 20;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_row_mask_ctrl, offset, bits_width, row_mask);
+    return;
+}
+
+void KEYSCAN_SetInColMask(uint32_t col_mask)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 20;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_col_mask_ctrl, offset, bits_width, col_mask);
+    return;
+}
+
+void KEYSCAN_SetIntTrigEn(uint8_t enable)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 1;
+    uint32_t data = (enable == 0) ? 0: 1;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_int_en, offset, bits_width, data);
+    return;
+}
+
+void KEYSCAN_SetFifoClrReg(uint8_t set)
+{
+    uint8_t offset = 4;
+    uint8_t bits_width = 1;
+    uint32_t data = (set == 0) ? 0: 1;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_int_en, offset, bits_width, data);
+    return;
+}
+
+uint8_t KEYSCAN_GetIntStateFifoEmptyRaw(void)
+{
+    uint8_t offset = 6;
+    uint8_t bits_width = 1;
+    uint8_t ret;
+
+    ret = KEYSCAN_reg_read_bits(&APB_KEYSCAN->key_int, offset, bits_width);
+    return ret;
+}
+
+uint8_t KEYSCAN_GetKeyScannerBusy(void)
+{
+    uint8_t offset = 6;
+    uint8_t bits_width = 1;
+    uint8_t ret;
+
+    ret = KEYSCAN_reg_read_bits(&APB_KEYSCAN->fifo_status, offset, bits_width);
+    return ret;
+}
+
+uint16_t KEYSCAN_GetKeyData(void)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 11;
+    uint16_t ret;
+
+    ret = KEYSCAN_reg_read_bits(&APB_KEYSCAN->key_data, offset, bits_width);
+    return ret;
+}
+
+void KEYSCAN_SetFifoNumTrigInt(uint32_t trig_num)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 5;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_trig, offset, bits_width, trig_num);
+    return;
+}
+
+void KEYSCAN_SetDmaNumTrigInt(uint32_t trig_num)
+{
+    uint8_t offset = 5;
+    uint8_t bits_width = 5;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_trig, offset, bits_width, trig_num);
+    return;
+}
+
+static uint8_t key_scan_row_to_idx[KEY_OUT_ROW_NUMBER] = {0};
+static uint8_t key_scan_col_to_idx[KEY_IN_COL_NUMBER]  = {0};
```

Comment:
move these into a `struct`, and allocated by developers:

```c
struct keyscan_ctx
{
    uint8_t key_scan_row_to_idx[KEY_OUT_ROW_NUMBER];
    uint8_t key_scan_col_to_idx[KEY_IN_COL_NUMBER]
};
```

Add a param `struct keyscan_ctx *p` to  `KEYSCAN_InitKeyScanToIdx`, etc.